### PR TITLE
feat: add normed 16 bit texture formats

### PR DESCRIPTION
**Pull request by pohlt**
Of course, the final PR will not change index.html.

Still missing (feel free to add):
* [x] RG and RGB formats
* [x] documentation (none required)

Discussion: https://forum.babylonjs.com/t/normalized-integer-textures-do-not-work/30333

WebGPU support seems to be [in the works](https://github.com/gpuweb/gpuweb/issues/3001) .

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

File: packages/tools/babylonServer/public/index.html [hunk 1]
```diff
@@ -1,43 +1,133 @@
 <!doctype html>
```
Comment (deltakosh) [inline packages/tools/babylonServer/public/index.html:1]: Please do not push that file

Everything else is fine

Comment (pohlt) [reply packages/tools/babylonServer/public/index.html:1]: Sure, it's actually the first sentence in the PR comment. 😉

I just included it here so a reviewer could easily test the functionality.

**Comment by pohlt** [comment]
Any guidance about documentation, tests, ...?

If not, I'm just going to add the RG and RGB flavors and remove the index.html.

File: packages/dev/core/src/Engines/engineCapabilities.ts [hunk 2]
```diff
@@ -121,4 +121,6 @@ export interface EngineCapabilities {
     supportFloatTexturesResolve: boolean;
     /** Defines if RG11B10UFloat texture format is color renderable */
     rg11b10ufColorRenderable: boolean;
+    /** Defines if EXT_texture_norm16 is available */
```
Comment (Popov72) [inline packages/dev/core/src/Engines/engineCapabilities.ts:124]: I think it's better to list the new supported formats when the boolean is true than giving the name of the extension, as the extension is WebGL specific.

File: packages/tools/babylonServer/public/index.html (hunk 1 above)
Comment (Popov72) [reply packages/tools/babylonServer/public/index.html:1]: In fact, you can create a regular PG to test the new feature, as we can test it with the PR merged thanks to a temporary link generated by the CI.

**Review by Popov72 — COMMENTED** [review]
That's great!

It would be nice to support all formats supported by the extension, maybe in a follow up?

**Comment by Popov72** [comment]
> Any guidance about documentation, tests, ...?
> 
> If not, I'm just going to add the RG and RGB flavors and remove the index.html.

Ah, I see you already planned to implement RG and RGB, so don't pay attention to my earlier comment about a follow up!

I don't think we really need new documentation or test for that, cc @sebavan just to be sure.

**Comment by Popov72** [comment]
The build error is because you must also set the new `textureNorm16` caps for the null, native and WebGPU engines. See files `Engines/nullEngine.ts`, `Engines/nativeEngine.ts` and `Engines/webgpuEngine.ts` (just set the boolean to false in these engines).

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15895/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15895/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15895/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15895/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15895/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15895/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15895/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15895/merge#BCU1XR#0

File: packages/dev/core/src/Engines/engineCapabilities.ts (hunk 2 above)
Comment (pohlt) [reply packages/dev/core/src/Engines/engineCapabilities.ts]: Ok, I added the new texture formats to the extension name, so if someone searches for it, she would find it in either case.

**Comment by pohlt** [comment]
> The build error is because you must also set the new `textureNorm16` caps for the null, native and WebGPU engines. See files `Engines/nullEngine.ts`, `Engines/nativeEngine.ts` and `Engines/webgpuEngine.ts` (just set the boolean to false in these engines).

Thanks for the pointers. Should be fixed now. Let's see what the CI run thinks...

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15895/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15895/merge/testResults/webgpuplaywright/index.html

**Review by Popov72 — APPROVED** [review]

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15895/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15895/merge/testResults/webgl2playwright/index.html

**Comment by pohlt** [comment]
[Here's a PG testing all eight flavors.](https://playground.babylonjs.com/?snapshot=refs/pull/15895/merge#SOCPNB#1) LGTM.

**Comment by Popov72** [comment]
> [Here's a PG testing all eight flavors.](https://playground.babylonjs.com/?snapshot=refs/pull/15895/merge#SOCPNB#2) 

Nice!

I've taken the liberty of updating it so that it also works with WebGPU when these formats are supported:

https://playground.babylonjs.com/?snapshot=refs/pull/15895/merge#SOCPNB#3

I forgot to comment, but you can remove the RGB formats from the WebGPU implementation, as they won't be supported.

Also, I think you can add it to the visualization tests (`tools/tests/test/visualization/config.json`), something like:
```json
        {
            "title": "Normed 16 bits texture formats",
            "playgroundId": "#SOCPNB#3",
            "excludedEngines": ["webgl1","webgpu"]
        },
```

When added to `config.json`, you can start the test by launching `npm run test:visualization:ui -- --project="webgl2"` on the command line. Choose the test and start it. As the reference image doesn't exist yet, it will be created.

**Comment by pohlt** [comment]
> I've taken the liberty of updating it so that it also works with WebGPU when these formats are supported:
> 
> https://playground.babylonjs.com/?snapshot=refs/pull/15895/merge#SOCPNB#7

Great, thanks! I replaced the `Math.random' with a basic RNG to make the results reproducible (for testing).
See https://playground.babylonjs.com/?snapshot=refs/pull/15895/merge#SOCPNB#8

> I forgot to comment, but you can remove the RGB formats from the WebGPU implementation, as they won't be supported.

Thanks for the input. I removed all references to RGB16* in WebGPU-related files.

> Also, I think you can add it to the visualization tests (`tools/tests/test/visualization/config.json`), something like:
> 
> ```json
>         {
>             "title": "Normed 16 bits texture formats",
>             "playgroundId": "#SOCPNB#7",
>             "excludedEngines": ["webgl1","webgpu"]
>         },
> ```
> 
> When added to `config.json`, you can start the test by launching `npm run test:visualization:ui -- --project="webgl2"` on the command line. Choose the test and start it. As the reference image doesn't exist yet, it will be created.

Ok, will give this a try.

**Comment by Popov72** [comment]
> Great, thanks! I replaced the `Math.random' with a basic RNG to make the results reproducible (for testing).

You didn't have too because the visu test framework already replaces `Math.random` with a reproducible RNG :).

I don't know how to add files to an existing PR, so here's the picture:

Normed-16-bits-texture-formats

It must be copied to `packages/tools/tests/test/visualization/ReferenceImages/` with name `Normed-16-bits-texture-formats.png`.

Also, I changed the position of the camera a bit, so the revision of the PG is `#SOCPNB#10`.

**Review by Popov72 — APPROVED** [review]

**Review by sebavan — APPROVED** [review]
Amazing team work !!!

**Pull request merged**